### PR TITLE
Issue #2702159: Added hook_mail for rules module.

### PR DESCRIPTION
**Pull request by yanniboi**
This is a PR for this issue: https://www.drupal.org/node/2702159

### CODE REVIEW

**Comment by yanniboi** [comment]
Tests coming...

File: tests/src/Kernel/RulesMailTest.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      ->setContextValue('subject', 'subject')
+      ->setContextValue('message', 'hello');
+
+    $params = [
+      'subject' => 'subject',
+      'message' => 'hello',
+    ];
+
+    // Send email.
+    \Drupal::service('plugin.manager.mail')->mail(
```
Comment (klausi) [inline tests/src/Kernel/RulesMailTest.php]: Not you should send the message, but the action. You should execute the action and then make sure the mail went out.

File: tests/src/Kernel/RulesMailTest.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  public function setUp() {
+    parent::setUp();
+    // Use the state system collector mail backend.
+    $this->container->get('config.factory')->getEditable('system.mail')
+      ->set('interface.default', 'test_mail_collector')
+      ->save();
+
+    // Reset the state variable that holds sent messages.
+    \Drupal::state()->set('system.test_mail_collector', []);
```
Comment (klausi) [inline tests/src/Kernel/RulesMailTest.php]: don't use \Drupal in classes, use $this->container instead.

File: tests/src/Kernel/RulesMailTest.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    \Drupal::service('plugin.manager.mail')->mail(
+      'rules', 'rules_action_mail_' . $action->getPluginId(),
+      $to[0],
+      LanguageInterface::LANGCODE_SITE_DEFAULT,
+      $params,
+      NULL
+    );
+
+    // Retrieve sent message.
+    $captured_emails = \Drupal::state()->get('system.test_mail_collector');
```
Comment (klausi) [inline tests/src/Kernel/RulesMailTest.php]: don't use \Drupal in classes, use $this->container instead.

File: tests/src/Kernel/RulesMailTest.php (hunk 1 above)
Comment (yanniboi) [inline tests/src/Kernel/RulesMailTest.php]: Ok cool

**Comment by yanniboi** [comment]
Fixed recommendations.

File: tests/src/Kernel/RulesMailTest.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  public function testSubjectAndBody() {
+    // Create action to send email.
+    $action = $this->actionManager->createInstance('rules_send_email');
+
+    $params = [
+      'to' => ['mail@example.com'],
+      'subject' => 'subject',
+      'message' => 'hello',
+    ];
```
Comment (klausi) [inline tests/src/Kernel/RulesMailTest.php:53]: Now that $params variable is not needed anymore. Just use the literals when setting the context values and when asserting the outcome.

**Comment by klausi** [comment]
merged, thanks!

**Pull request closed without merging**